vllm/lora/punica_wrapper/punica_base.py

```diff
@@ -0,0 +1,480 @@
+    """
+
+    def __init__(self, max_num_batched_tokens: int, max_batches: int,
+                 device: Union[torch.device, str], **kwargs):
+        self._token_lora_indices = torch.empty(max_num_batched_tokens,
+                                               dtype=torch.long,
+                                               device=device)
+        self._sampler_indices = torch.empty(max_num_batched_tokens,
+                                            dtype=torch.long,
+                                            device=device)
+        self._sampler_indices_padded = torch.empty(max_num_batched_tokens,
+                                                   dtype=torch.long,
+                                                   device=device)
+        self._embeddings_indices = torch.empty(2,
+                                               max_num_batched_tokens,
+                                               dtype=torch.long,
+                                               device=device)
+        self._long_lora_indices = torch.empty(max_num_batched_tokens,
+                                              dtype=torch.long,
+                                              device=device)
+
+        # 5 is the number of indicies tensors.
+        # base_indices, sampler_indices, sampler_indices_padded,
+        # embeddings_indices,long_lora_indices
+        self.indices_len: List[Optional[int]] = [None] * 5
+        # these attributes are the information required for sgmv kernel
+        self._seq_start_locs = torch.empty(max_batches,
+                                           dtype=torch.long,
+                                           device=device)
+        self._seq_lengths = torch.empty(max_batches,
+                                        dtype=torch.long,
+                                        device=device)
+        self._lora_indices_per_batch = torch.empty(max_batches,
+                                                   dtype=torch.long,
+                                                   device=device)
+        self.device: torch.device = device
+        self.max_length: int = 0
+        self.token_nums: int = 0
+        self.batch_size: int = -1
+        self.is_prefill = False
+        self.no_lora = False
+
+    def _update_base_metadata(
+        self,
+        mapping: "LoRAMapping",
+        lora_index_to_id: List[Optional[int]],
+        max_loras: int,
+        vocab_size: int,
+        extra_vocab_size: int,
+        long_lora_context: Optional["LongContextLoRAContext"] = None,
+    ):
+        (
+            base_indices,
+            sampler_indices,
+            sampler_indices_padded,
+            embeddings_indices,
+            long_lora_offsets_tensor,
+            indices_len,
+        ) = convert_mapping(
+            mapping,
+            lora_index_to_id,
+            max_loras,
+            vocab_size,
+            extra_vocab_size,
+            self.device,
+            long_lora_context,
+        )
```

Comment:
Perhaps we can implement a `PunicaMetadata` class as well to seperate metadata handling helper functions out of PunicaWrapper to make it cleaner? (Just like what we do in AttentionMeta)

Reply:
make sense

Reply:
I'll try it next week. Is there anything else that needs to be modified? @Isotr0py  thanks for your view

Reply:
The others LGTM :)

Reply:
After some consideration, I  want to implement your requested improvements in following  PRs, mainly because:
- We are currently working on implementing full-pipeline LoRA support for VL (Vision-Language), and we're uncertain whether additional metadata will be needed.

Reply:
After some consideration, I  want to implement your requested changes in following  PR, mainly because:
- We are currently working on implementing full-pipeline LoRA support for VL model,  and we're uncertain whether additional metadata will be needed.
- During the prefill phase, these metadata can actually be obtained from the attention metadata. We are also considering how to directly utilize the attention metadata -  like [get_forward_context](https://github.com/vllm-project/vllm/blob/main/vllm/attention/layer.py#L243) might be feasible as well
- We are also exploring triton kernel fusion.

Therefore, I suggest implementing these changes in subsequent PRs. However, if you feel they should be completed in this PR, I have no objections.

Reply:
No problem. Let's remain it to following PR!